src/core/global.c

```diff
@@ -1465,3 +1483,69 @@ void nn_global_rele_socket(struct nn_sock *sock)
     nn_sock_rele(sock);
     nn_glock_unlock();
 }
+
+int nn_setopt (int option, const void *optval, size_t optvallen)
+{
+    int idx;
+
+    switch (option) {
+    case NN_FORK_STRATEGY:
+        nn_assert(optvallen == sizeof (int));
+        idx = *(const int *) optval;
+        if (idx < 0 || idx >= NN_FORK_MAX_) {
+            errno = EINVAL;
+            break;
+        }
+#ifndef NN_HAVE_WINDOWS
+        nn_fork_strategy = &nn_fork_strategies[idx];
+#endif
+        return 0;
+    default:
+        errno = ENOPROTOOPT;
+        break;
+    }
+    return -1;
+}
+
+int nn_global_lock_all_sockets (void)
+{
+    int i;
+
+    if (self.socks && self.nsocks) {
+        for (i = 0; i != NN_MAX_SOCKETS; ++i)
+            if (self.socks [i])
+                nn_ctx_enter (nn_sock_getctx (self.socks [i]));
+    }
+}
```

Comment:
> the global lock is already held

Indeed.

> make it so this function explicitely calls `nn_glock_lock()` then `nn_glock_unlock()`

Perhaps not, if the semantic required here is to remain locked until the symmetric work is done also, or if there is additional initial work that needs to be done along with the call to this. That said; your mind is better wrapped around what is required right now; consider my comments at this point "drive by" considerations.

> word of warning that the global lock needs to be held

Comments alone are weak; thinking aloud, it would be perfect if there could be created a macro `nn_assert_global_lock_held();` to capture this thought in code. That could be used a few different places, probably (notably ... pretty much anywhere the global lock is held, but *especially* in locations where the `unlock` does not immediately follow the `lock` in the same code block/scope)